### ISSUE TITLE
error messages incorrectly display where bounds with higher ranked self types

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#104398]
```rust
struct Inv<'a>(&'a mut &'a ());

trait Trait {}

fn with_bound<T>()
where
    (for<'a> fn(Inv<'a>)): Trait,
{
}

fn with_diag_bound<T>()
where
    for<'a> fn(Inv<'a>): Trait,
{
}

fn main() {
    with_bound();
}
```

The current output is:

```
warning: unnecessary parentheses around type
 --> src/main.rs:7:5
  |
7 |     (for<'a> fn(Inv<'a>)): Trait,
  |     ^                   ^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
7 -     (for<'a> fn(Inv<'a>)): Trait,
7 +     for<'a> fn(Inv<'a>): Trait,
  |

error[[E0277]](https://doc.rust-lang.org/stable/error-index.html#E0277): the trait bound `for<'a> fn(Inv<'a>): Trait` is not satisfied
  --> src/main.rs:18:5
   |
18 |     with_bound();
   |     ^^^^^^^^^^ the trait `Trait` is not implemented for `for<'a> fn(Inv<'a>)`
   |
note: required by a bound in `with_bound`
  --> src/main.rs:7:28
   |
5  | fn with_bound<T>()
   |    ---------- required by a bound in this
6  | where
7  |     (for<'a> fn(Inv<'a>)): Trait,
   |                            ^^^^^ required by this bound in `with_bound`
```
The incorrect warning is tracked in #104397. The error message is also wrong though. `for<'a> fn(Inv<'a>): Trait` is a different from `(for<'a> fn(Inv<'a>)): Trait`.

**Comment by estebank** [comment]
Current output:

```
error[E0277]: the trait bound `for<'a> fn(Inv<'a>): Trait` is not satisfied
  --> src/main.rs:12:5
   |
12 |     with_bound();
   |     ^^^^^^^^^^^^ the trait `Trait` is not implemented for `for<'a> fn(Inv<'a>)`
   |
help: this trait has no implementations, consider adding one
  --> src/main.rs:3:1
   |
3  | trait Trait {}
   | ^^^^^^^^^^^
note: required by a bound in `with_bound`
  --> src/main.rs:7:28
   |
5  | fn with_bound<T>()
   |    ---------- required by a bound in this function
6  | where
7  |     (for<'a> fn(Inv<'a>)): Trait,
   |                            ^^^^^ required by this bound in `with_bound`
```
